### PR TITLE
[mlir][llvm] Drop unreachable basic block during import

### PR DESCRIPTION
**Pull request by gysit**
This revision updates the LLVM IR import to support unreachable basic blocks. An unreachable block may dominate itself and a value defined inside the block may thus be used before its definition. The import does not support such dependencies. We thus delete the unreachable basic blocks before the import. This is possible since MLIR does not have basic block labels that can be reached using an indirect call and unreachable blocks can indeed be deleted safely.

Additionally, add a small poison constant import test.

### CODE REVIEW

**Comment by joker-eph** [comment]
I'm confused because the title says "Import unreachable basic blocks" but the description says "We thus delete the unreachable basic block"?

**Comment by gysit** [comment]
> I'm confused because the title says "Import unreachable basic blocks" but the description says "We thus delete the unreachable basic block"?

I tried to improve title and comment. That was indeed a failed attempt of coming up with a short title.

The revision indeed deletes unreachable basic blocks before the import since importing an unreachable block is not possible in the general case. The problem is that instructions can use a value that they define themselves. This is possible since an unreachable block can dominate itself.

The fadd in the following code is an example for such a self dependency.
```
define void @unreachable_block(float %0) {
.entry:
  ret void

unreachable:
  %1 = fadd float %0, %1
  br label %unreachable
}
```

**Comment by joker-eph** [comment]
> The revision indeed deletes unreachable basic blocks before the import since importing an unreachable block is not possible in the general case. The problem is that instructions can use a value that they define themselves. This is possible since an unreachable block can dominate itself.

I know about self-referencing instructions, but I don't quite get why we can't import this in MLIR?

**Comment by Dinistro** [comment]
Turns out the MLIR also supports this. I was honestly not aware of this, and I fail to see any usecase for this, as the conversion framework doesn't touch unreachable blocks. It seems adding support for that might be possible by relaxing some assumptions, but the canonicalizer will afterwards just remove these unreachable blocks.

Thus, I'm in favor of taking the path of least resistance:
If this can be done easily, lets import these self-referencing operations, otherwise drop these constructs directly in the import, as done in this PR.

**Comment by gysit** [comment]
> I know about self-referencing instructions, but I don't quite get why we can't import this in MLIR?

The import in its current form cannot support this since all operand values need to exist before the operation is converted. There is currently no mechanism to create temporary values that then later on are magically replaced. A possibly solution may be to have an LLVM pass before the import that introduces phi instructions for such values. That sounds really complex to get right in the general case though. 

Is there a use case for such unreachable blocks in MLIR? Canonicalize immediately deletes such a block since there to the best of my knowledge there is no way to reach an unreachable block in MLIR (since there is no indirect branch or similar). I would thus strongly favor the current solution which is much simpler and leads to the same result.

File: mlir/test/Target/LLVMIR/Import/constant.ll [hunk 1]
```diff
@@ -47,6 +47,16 @@ define void @undef_constant(i32 %arg0) {
 
 ; // -----
 
+; CHECK-LABEL: @poison_constant
+define void @poison_constant(double %arg0) {
```
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/Import/constant.ll:51]: Nit: This is not really related to the remainder of this PR. Still, we did not test this before, so it might make sense to keep it here?

File: mlir/test/Target/LLVMIR/Import/unreachable-blocks.ll
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/Import/unreachable-blocks.ll:1]: Maybe also add a test where the unreachable block branches to another reachable one. To ensure that the graph traversal isn't some connected component like thing.
I wonder what happens if there are phi-nodes involved that point to blocks that we will drop during the import.

File: mlir/test/Target/LLVMIR/Import/constant.ll (hunk 1 above)
Comment (gysit) [reply mlir/test/Target/LLVMIR/Import/constant.ll:51]: Yes I added it intentionally since it was missing.

File: mlir/test/Target/LLVMIR/Import/unreachable-blocks.ll
Comment (gysit) [reply mlir/test/Target/LLVMIR/Import/unreachable-blocks.ll:1]: Good point! I added a back edge with a value. The Phi-nodes are fine since the block references are only used when translating the branch instruction in the predecessor block. This code never executes for the unreachable block.

**Comment by joker-eph** [comment]
> The import in its current form cannot support this since all operand values need to exist before the operation is converted. There is currently no mechanism to create temporary values that then later on are magically replaced. 

Sure, but such mechanism could be implemented fairly easily right?

> Is there a use case for such unreachable blocks in MLIR? 

There may not be actual use-cases to import them in practice. It just makes me always a bit nervous when the translation isn't "faithful" ;)

**Comment by gysit** [comment]
> Sure, but such mechanism could be implemented fairly easily right?

It is definitely more complex than the current solution. It requires at least one additional mapping from the initially unknown values to all their uses and some dummy poison values that are used initially. Plus I guess we can only check at the end of the translation if all unknown values have been set. If possible I would like to avoid that complexity.

**Comment by gysit** [comment]
I prototyped a version that imports the code as is without deleting the unreachable blocks. It turns out to be quite tricky to get right since the change interferes with the error handling of the import. In particular, the `convertValue` method is used in different contexts and only in some of them it is ok to introduce a placeholder value. After the import, the unreachable blocks either way need to be deleted since the conversion infrastructure does not support them AFAIK.

As this PR fixes an existing bug, I suggest to land it (it is definitely an improvement over the status quo). Once we have a concrete use case, we can follow up with an version that introduces placeholder variables to support self-referencing instructions.

**Review by Dinistro — APPROVED** [review]
Thanks for addressing the comments. I'm fine with fixing this bug and deferring a potential support for unreachable basic blocks to a later revision.
